students/paul_jurek/lesson05/mailroom.py

```diff
@@ -0,0 +1,178 @@
+    sorted by total historical donation amount.
+    Include Donor Name, total donated, number of donations and average
+    donation amount as values in each row. You do not need to print out all
+    their donations, just the summary info.
+    Using string formatting, format the output rows as nicely as possible.
+    The end result should be tabular (values in each column should align
+    with those above and below)
+    After printing this report, return to the original prompt.
+    At any point, the user should be able to quit their current task and
+    return to the original prompt.
+    From the original prompt, the user should be able to quit the script
+    cleanly.
+    Your report should look something like this:
+    Donor Name                | Total Given | Num Gifts | Average Gift
+    ------------------------------------------------------------------
+    William Gates, III         $  653784.49           2  $   326892.24
+    Mark Zuckerberg            $   16396.10           3  $     5465.37
+    Jeff Bezos                 $     877.33           1  $      877.33
+    Paul Allen                 $     708.42           3  $      236.14
+    """
+    print(f"{'Donor Name':<26}|{'Total Given':^15}|"
+          f"{'Num Gifts':^11}|{'Average Gift':^15}")
+    print('-'*70)
+    donor_stats = []
+    for donor in donors.keys():
+        donor_stats.append(summarize_donor(donor))
+
+    donor_stats.sort(key=lambda tup: tup[1], reverse=True)
+    for summary in donor_stats:
+        print(f"{summary[0]:<26} ${summary[1]:>13.2f}  "
+              f"{summary[2]:>10}  ${summary[3]:>14.2f}")
+
+
+def summarize_donor(donor_name):
+    """generates donor summary
+    args:
+        donor_name: donor name matching key from donors
+    returns
+        tuple with fields (donor name, total given, num gifts, average gift)"""
+    total_given = sum(donors[donor_name])
+    num_gifts = len(donors[donor_name])
+    average_gift = total_given/num_gifts
+
+    return(donor_name, total_given, num_gifts, average_gift)
+
+
+def send_letters_to_everyone():
+    """process to evaluate all donors and create letter to send to
+    donors."""
+    # iterate through donors and donations to send thank yous
+    THANK_YOU_DIRECTORY = '/mailroom_thankyou_letters/'
+    for donor in donors:
+        file_name = "".join([THANK_YOU_DIRECTORY,
+                             donor.replace(" ", "_").lower(), '.txt'])
+        donor_info = summarize_donor(donor)
+        thank_you_text = create_donation_thank_you(fullname=donor,
+                                                   amount=donor_info[1])
+        try:
+            with open(file_name, 'w') as f:
+                f.write(thank_you_text)
+        except FileNotFoundError:
+            print('Mailroom thank you directory not found.  Please create this directory first.')
+            break
+    print(f'Thank you letters create in "{THANK_YOU_DIRECTORY}"')
```

Comment:
i think this should go into `else` block of try/except? since this should only print if operation was successful.